### PR TITLE
Analytics Implementation for Schematics and Command Flags

### PR DESCRIPTION
**Pull request by hansl**
This does not include analytics for project build and rebuilds yet. It will be in a follow up PR.

### CODE REVIEW

**Comment by hansl** [comment]
Need https://github.com/angular/angular/pull/29382 to go in first.

**Comment by hansl** [comment]
@jenniferfell For the messaging in `packages/angular/cli/models/analytics.ts`.

File: docs/design/analytics.md [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+### List Of All Dimensions
+| Id | Category | Flag | Type | File / Description |
+|:---:|:---|:---|:---|---:|
+| 1 | `generate`, `new` | `--dryRun`      | `Boolean` | |
+| 2 | `generate`, `new` | `--force`       | `Boolean` | |
+| 3 | `generate`, `new` | `--interactive` | `Boolean` | |
+| 4 | `generate` | `--skipInstall` | `Boolean` | |
+| 5 | `generate` | `--style` | `String` | |
+| 6 | `add` | `--collection` | `String` | Only for packages that we control (see safelist in `add-impl.ts`). |
+| 7 | `build`, `serve` | `--buildEventLog` | `Boolean` | If the flag was used (does not report its value). See `build-impl.ts`. |
```
[the remark this answers is not in the document]
Comment (hansl) [reply docs/design/analytics.md]: I added the `optimization` and `aot` keys to the dimensions, so we'll track those.

File: docs/design/analytics.md [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+# Debugging
+Using `DEBUG=universal-analytics` will report all calls to the universal-analytics library,
+including queuing events and sending them to the server.
+
+Using `DEBUG=ng:analytics` will report additional information regarding initialization and
+decisions made during the usage analytics process, e.g. if the user has analytics disabled.
+
+Using `DEBUG=ng:analytics:command` will show the decisions made by the command runner.
+
+Using `DEBUG=ng:analytics:log` will show what we actually send to GA.
```
[the remark this answers is not in the document]
Comment (hansl) [reply docs/design/analytics.md]: Wish list, I created a story for this: https://angular-team.atlassian.net/browse/TOOL-762

File: etc/api/angular_devkit/schematics/tools/index.d.ts [hunk 3]
```diff
@@ -47,7 +51,7 @@ export declare class FileSystemEngineHost extends FileSystemEngineHostBase {
     hasTaskExecutor(name: string): boolean;
 }
 
-export declare abstract class FileSystemEngineHostBase implements EngineHost<FileSystemCollectionDescription, FileSystemSchematicDescription> {
+export declare abstract class FileSystemEngineHostBase implements FileSystemEngineHost {
```
[the remark this answers is not in the document]
Comment (hansl) [reply etc/api/angular_devkit/schematics/tools/index.d.ts]: Agreed, a lot of the OG schematics files were made without much comment. I can add some here as I go.

File: docs/design/analytics.md [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+See [the `debug` NPM library](https://www.npmjs.com/package/debug) for more information.
+
+# Disabling Telemetry
+There are 2 ways of disabling usage analytics:
+
+1. using `ng analytics off` (or changing the global configuration file yourself). This is the same
```
[the remark this answers is not in the document]
Comment (hansl) [reply docs/design/analytics.md]: Yes, and `ng config` isn't quite the same. Also it's better to communicate intent to the user to have its own command with its own documentation and help section.

**Comment by alexeagle** [comment]
The `x-user-analytics` bits are sprinkled around, and should have a sequence - I know it's documented in the readme, but maybe we should have a golden file test to verify later changes?
The readme content could be the golden file

File: packages/angular/cli/commands/add-impl.ts [hunk 5]
```diff
@@ -23,6 +24,14 @@ import { Schema as AddCommandSchema } from './add';
 
 const npa = require('npm-package-arg');
 
+const safelistNgAddCollectionAnalytics = [
+  /^@angular\//,  // All @angular packages are owned by us.
+  /^@angular-devkit\//,  // All @angular-devkit packages are owned by us.
+  '@schematics/angular',
```
[the remark this answers is not in the document]
Comment (hansl) [reply packages/angular/cli/commands/add-impl.ts]: That was always the plan. Nobody worked with us for it, and the few people that we talked to didn't show much interest (NativeScript would rather have it in their own space, for example), but better be conservative on safelist.

File: etc/api/angular_devkit/schematics/tools/index.d.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (hansl) [reply etc/api/angular_devkit/schematics/tools/index.d.ts]: There are comments, but comments are (somehow) dropped when generating the API golden files. Maybe we should consider comments (which are shown in developer help for example) as part of our public API.

File: etc/api/angular_devkit/schematics/tools/index.d.ts [hunk 6]
```diff
@@ -131,7 +138,7 @@ export declare class NodeWorkflow extends workflow.BaseWorkflow {
     });
 }
 
-export declare type OptionTransform<T extends object, R extends object> = (schematic: FileSystemSchematicDescription, options: T, context?: FileSystemSchematicContext) => Observable<R>;
+export declare type OptionTransform<T extends object, R extends object> = (schematic: FileSystemSchematicDescription, options: T, context?: FileSystemSchematicContext) => Observable<R> | PromiseLike<R> | R;
```
[the remark this answers is not in the document]
Comment (hansl) [reply etc/api/angular_devkit/schematics/tools/index.d.ts]: A function that matches `OptionTransform` is given to us, so it's about being liberal in what we accept as input.

**Comment by hansl** [comment]
I'm adding a validation script to check that `x-user-analytics` are properly kept in sync with the `analytics.md` file.

File: docs/design/analytics.md [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+# Pageview
+Each command creates a pageview with the path `/command/${commandName}/${subcommandName}`. IE.
+`ng generate component my-component --dryRun` would create a page view with the path
+`/command/generate/@schematics_angular/component`.
+
+Additionally, only the Architect configuration of `default`, `production` and `staging` will be
+logged. The project name will be removed, and if the configuration is different than the three
+basic ones, it will be replaced with `_`. 
+The command `ng run some-project:lint:some-configuration` will create a page view with the path
+`/command/run/_:lint:_`. Configurations are only used when actually mentionned on the command
```
Comment (alan-agius4) [inline docs/design/analytics.md]: mentionned -> mentioned

File: packages/angular/cli/commands/add-impl.ts [hunk 8]
```diff
@@ -139,6 +148,29 @@ export class AddCommand extends SchematicCommand<AddCommandSchema> {
...
+    function _matches(re: RegExp | string, str: string): boolean {
+      if (typeof re == 'string') {
+        return str === re;
+      } else {
+        return re.test(str);
+      }
+    }
+
+    // Add the collection if it's safe listed.
+    if (collection && safelistNgAddCollectionAnalytics.find(re => _matches(re, collection))) {
```
Comment (alan-agius4) [inline packages/angular/cli/commands/add-impl.ts]: nit: `.find` -> `.some`

File: docs/design/analytics.md [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+Using `DEBUG=ng:analytics:log` will show what we actually send to GA.
+
+See [the `debug` NPM library](https://www.npmjs.com/package/debug) for more information.
+
+# Disabling Telemetry
+There are 2 ways of disabling usage analytics:
+
+1. using `ng analytics off` (or changing the global configuration file yourself). This is the same
+   as answering "No" to the prompt.
+1. There is an `NG_CLI_TELEMETRY` environment variable that overrides the global configuration.
```
Comment (alan-agius4) [inline docs/design/analytics.md]: Cannot find any code referencing `NG_CLI_TELEMETRY`. Am I missing something?

File: packages/angular/cli/commands/config-impl.ts [hunk 10]
```diff
@@ -32,6 +32,9 @@ const validCliPaths = new Map([
   ['cli.warnings.versionMismatch', 'boolean'],
   ['cli.defaultCollection', 'string'],
   ['cli.packageManager', 'string'],
+  ['cli.analytics', 'string'],
```
Comment (alan-agius4) [inline packages/angular/cli/commands/config-impl.ts]: Why are these being used?

File: docs/design/analytics.md (hunk 7 above)
Comment (hansl) [reply docs/design/analytics.md]: Done.

File: packages/angular/cli/commands/add-impl.ts (hunk 8 above)
Comment (hansl) [reply packages/angular/cli/commands/add-impl.ts]: Using a call to `isPackageNameSafeForAnalytics(collection)` instead.

File: docs/design/analytics.md (hunk 9 above)
Comment (hansl) [reply docs/design/analytics.md]: It's `NG_CLI_ANALYTICS` now. Renamed in the md.

File: packages/angular/cli/commands/config-impl.ts (hunk 10 above)
Comment (hansl) [reply packages/angular/cli/commands/config-impl.ts]: `ng config cli.analytics=abc`. I think the whole config command is a bit weird in its type checking, but it's not the right time to refactor that one yet.

**Pull request merged**